Add semantic conventions for instrumenting AWS Lambda. (#1442)
open-telemetry/opentelemetry-specification: specification/trace/semantic_conventions/instrumentation/aws-lambda.md

> # Instrumenting AWS Lambda

Reviewer: This document doesn't mention its relationship to the FaaS spec, https://github.com/open-telemetry/opentelemetry-specification/blob/main/semantic_conventions/trace/faas.yaml. Is it extending it or overriding it?

Author: Doesn't the first paragraph clarify this, including an actual link to faas? Let me know if something's unclear.

**Status**: [Experimental](../../../document-status.md)

> This document defines how to fill semantic conventions when instrumenting an AWS Lambda request handler. AWS

Reviewer 2: ```suggestion
This document defines how to apply semantic conventions when instrumenting an AWS Lambda request handler. AWS
```

Lambda largely follows the conventions for [FaaS](../faas.md) while [HTTP](../http.md) conventions are also
applicable when handlers are for HTTP requests.

There are a variety of triggers for Lambda functions, and this document will grow over time to cover all the
use cases.

## All triggers

> For all events, a span with kind `SERVER` MUST be created corresponding to the function invocation. Unless

Reviewer 2: ```suggestion
For all events, a span with kind `SERVER` MUST be created corresponding to the function invocation unless stated otherwise below. Unless
```

SQS wants CONSUMER

Author: We generate two additional spans in addition for SQS. IIRC this was to have a span that contains the function name.

I notice after reading through the code to write this that for HTTP we actually are overwriting the name of the top span instead of generating a new one. I guess this should make sense for SQS as well, I can reflect that (and we'd need to update the Java).

Author: @Oberon00 Updated to only have two SQS spans. @kubawach can you confirm this should be ok?

stated otherwise below, the name of the span MUST be set to the function name from the Lambda `Context`.

<!-- semconv aws.lambda -->
| Attribute  | Type | Description  | Examples  | Required |
|---|---|---|---|---|

> | [`cloud.account.id`](../../../resource/semantic_conventions/cloud.md) | string | The account ID for the function. If not provided on Lambda `Context`, it SHOULD be parsed from the value of `faas.id` as the fifth item when splitting on `:`. | `opentelemetry` | No |

Reviewer: A better example here is 123456789012.

| [`faas.execution`](../faas.md) | string | The value of the AWS Request ID from the Lambda `Context`. | `943ad105-7543-11e6-a9ac-65e093327849` | No |
| [`faas.id`](../../../resource/semantic_conventions/faas.md) | string | The value of the invocation arn for the function from the Lambda `Context`. [1] | `arn:aws:lambda:us-east-2:123456789012:function:my-function:1` | No |

> **[1]:** For example, in AWS Lambda this field corresponds to the [ARN](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html) value, in GCP to the URI of the resource, and in Azure to the [FunctionDirectory](https://github.com/Azure/azure-functions-host/wiki/Retrieving-information-about-the-currently-running-function) field.

Reviewer 2: That copied generic note is confusing. Does `note: ""` work in the YAML? Otherwise this would be worth a bugfix to the tool IMHO.

Author: Nope same issue in my other PR

<!-- endsemconv -->

The parent of the span MUST be determined by considering both the environment and any headers or attributes
available from the event.

> If the `_X_AMZN_TRACE_ID` environment variable is set, it SHOULD be parsed into an OpenTelemetry `Context` using

Reviewer 2: It seems to me that the application owner should be able to decide this. Ideally, the X-Ray propagator (or a specialized LambdaXrayPropagator) would be written in such a way that it checks the environment variable itself. The instrumentation should not decide this.

Author: The user can currently decide by enabling or disabling XRay so I think this reflects that choice well. If we had another option in the instrumentation it's two settings related to XRay which I think is just an extra setting.

It reminds me that in a follow-up to the SDK PR I need to describe using the AWS propagator directly as we do in Java. It's the only recognized format for the next years so there is no point for an option at least until that changes it's just an extra option. I think it's similar for Lambda.

the [AWS X-Ray Propagator](../../../context/api-propagators.md). If the resulting `Context` is sampled, then this
`Context` is the parent of the function span. The environment variable will be set and the `Context` will be
sampled only if AWS X-Ray has been enabled for the Lambda function. A user can disable AWS X-Ray for the function
if this propagation is not desired.

Otherwise, for an API Gateway Proxy Request, the user's configured propagators should be applied to the HTTP
headers of the request to extract a `Context`.

## API Gateway

API Gateway allows a user to trigger a Lambda function in response to HTTP requests. It can be configured to be
a pure proxy, where the information about the original HTTP request is passed to the Lambda function, or as a
configuration for a REST API, in which case only a deserialized body payload is available.  In the case the API
gateway is configured to proxy to the Lambda function, the instrumented request handler will have access to all
the information about the HTTP request in the form of an API Gateway Proxy Request Event.

The Lambda span name SHOULD be set to the `Resource` from the proxy request event, which corresponds to the user

> configured HTTP route instead of the function name.

Reviewer 2: Would it make sense to set `http.route` too?

Author: Yeah think so

<!-- semconv aws.lambda.api-gateway-proxy -->
| Attribute  | Type | Description  | Examples  | Required |
|---|---|---|---|---|

> | [`faas.trigger`](../faas.md) | string | MUST be `http`. | `http` | No |

Reviewer 2: Consider using `required: always` in the YAML.

Reviewer 2: This seems like a mistake to me, should faas.trigger really be optional here?

# Instrumenting AWS Lambda

**Status**: [Experimental](../../../document-status.md)

This document defines how to apply semantic conventions when instrumenting an AWS Lambda request handler. AWS
Lambda largely follows the conventions for [FaaS](../faas.md) while [HTTP](../http.md) conventions are also
applicable when handlers are for HTTP requests.

There are a variety of triggers for Lambda functions, and this document will grow over time to cover all the
use cases.

> 

Reviewer 3: ```suggestion

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->
// toc goes here
<!-- tocstop -->

```

A TOC would be nice now that the file ended up being a bit longer.

| Parent |  | Span Client |
| SpanKind | `CLIENT` | `SERVER` |
| Status | `Ok` | `Ok` |
| `faas.execution` | | `79104EXAMPLEB723` |
| `faas.id` | | `arn:aws:lambda:us-west-2:123456789012:function:my-function` |
| `faas.trigger` | | `http` |
| `cloud.account.id` | | `12345678912` |
| `net.peer.name` | `foo.execute-api.us-east-1.amazonaws.com` |  |
| `net.peer.port` | `413` |  |
| `http.method` | `GET` | `GET` |
| `http.user_agent` | `okhttp 3.0` | `okhttp 3.0` |
| `http.url` | `https://foo.execute-api.us-east-1.amazonaws.com/pets/10` |  |
| `http.scheme` | | `https` |
| `http.host` | | `foo.execute-api.us-east-1.amazonaws.com` |
| `http.target` | | `/pets/10` |
| `http.route` | | `/pets/{petId}` |
| `http.status_code` | `200` | `200` |

### API Gateway Request Proxy (Lambda tracing active)

Active tracing in Lambda means an API Gateway span `Span APIGW` and a Lambda runtime invocation span `Span Lambda`
will be exported to AWS X-Ray by the infrastructure (not instrumentation). All attributes above are the same
except that in this case, the parent of `APIGW` is `Span Client` and the parent of `Span Function` is
`Span Lambda`. This means the hierarchy looks like:

> Span Client --> Span APIGW --> Span Lambda --> Span Function

Reviewer 3: Please put this in a fenced code block (suggesting "```" here doesn't work apparently 😄).

In UML the arrows would be the other way round (`parent <-- child`). I'm not sure what would be most clear here.

Author: Added the fence. Kept the arrow direction, in tracing we're watching a request flow through steps, IMO having the order model the request direction is clearest.

### SQS (Lambda tracing passive)

Given a process P, that sends two messages to a queue Q on SQS, and a Lambda function F, which processes both of them in one batch (Span ProcBatch) and
generates a processing span for each message separately (Spans Proc1 and Proc2).

```
Process P: | Span Prod1 | Span Prod2 |
--
Function F:                      | Span ProcBatch |
                                        | Span Proc1 |
                                               | Span Proc2 |
```

| Field or Attribute | Span Prod1 | Span Prod2 | Span ProcBatch | Span Proc1 | Span Proc2 |
|-|-|-|-|-|-|
| Span name | `Q send` | `Q send` | `Q process` | `Q process` | `Q process` |
| Parent |  |  |  | Span ProcBatch | Span ProcBatch |
| Links |  |  |  | Span Prod1 | Span Prod2 |
| SpanKind | `PRODUCER` | `PRODUCER` | `CONSUMER` | `CONSUMER` | `CONSUMER` |
| Status | `Ok` | `Ok` | `Ok` | `Ok` | `Ok` |
| `messaging.system` | `AmazonSQS` | `AmazonSQS` | `AmazonSQS` | `AmazonSQS` | `AmazonSQS` |

> | `messaging.destination` | `Q` | `Q` | | `Q` | `Q` | `Q` |

Reviewer 3: ```suggestion
| `messaging.destination` | `Q` | `Q` | `Q` | `Q` | `Q` |
```
One column too many :-)

Author: I laughed that I ended up making the column length typo anyways!

| `messaging.destination_kind` | `queue` | `queue` | `queue` | `queue` | `queue` |
| `messaging.operation` |  |  | `process` | `process` | `process` |
| `messaging.message_id` | | | | `"a1"` | `"a2"` |

The above requires user code change to create `Span Proc1` and `Span Proc2`. In Java, the user would inherit from
[TracingSqsMessageHandler][] instead of Lambda's standard `RequestHandler` to enable them. Otherwise these two spans
would not exist.

[TracingSqsMessageHandler]: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/v1.0.1/instrumentation/aws-lambda-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambda/v1_0/TracingSqsMessageHandler.java

### SQS (Lambda tracing active)

Active tracing in Lambda means a Lambda runtime invocation span `Span Lambda` will be exported to X-Ray by the
infrastructure (not instrumentation). In this case, all of the above is the same except `Span ProcBatch` will
have a parent of `Span Lambda`. This means the hierarchy looks like:

Span Lambda --> Span ProcBatch --> Span Proc1 (links to Span Prod1 and Span Prod2)

>                                \-> Span Proc2 (links to Span Prod1 and Span Prod2)

Reviewer 3: Same as above (fences and arrow direction).

# Instrumenting AWS Lambda

**Status**: [Experimental](../../../document-status.md)

This document defines how to apply semantic conventions when instrumenting an AWS Lambda request handler. AWS
Lambda largely follows the conventions for [FaaS](../faas.md) while [HTTP](../http.md) conventions are also
applicable when handlers are for HTTP requests.

There are a variety of triggers for Lambda functions, and this document will grow over time to cover all the
use cases.

## All triggers

For all events, a span with kind `SERVER` MUST be created corresponding to the function invocation unless stated
otherwise below. Unless stated otherwise below, the name of the span MUST be set to the function name from the
Lambda `Context`.

> The following attributes SHOULD be set.

Reviewer 4: Does it make sense to use `MUST` for `faas.execution`/`faas.id`?  Asking because I'd like to better understand "instrumentation semconv" and how strict/lenient we're measuring compliance.

Author: Yeah I'm not too sure about this so open to suggestions. I generally went with `SHOULD` for attributes read from the instrumentation library mainly because for me to comfortably say `MUST`, I'd need to verify every single language's Lambda framework to make sure these actually are available. Perhaps I should do so, though I think for some frameworks this can be extremely tedious (many languages have 2-3 versions of the AWS SDK by now...), so leaning towards SHOULD indicating if they're available, fill them plz.

Reviewer 5: +1 for being more flexible here (number of SDKs x number of available instrumentations)

> - [`faas.execution`](../faas.md) - This is always available through an accessor on the Lambda `Context`

Reviewer 2: IIRC this should be set to the request ID?

Author: Oops, thanks!

- [`faas.id`](../../../resource/semantic_conventions/faas.md) - This is always available through an accessor on the Lambda `Context`
- [`cloud.account.id`](../../../resource/semantic_conventions/cloud.md) - In some languages, this is available as an accessor on the Lambda `Context`. Otherwise, it can be parsed from the value of `faas.id` as the fifth item when splitting on `:`

The parent of the span MUST be determined by considering both the environment and any headers or attributes

> available from the event.

Reviewer 2: ```suggestion
available from the event, as explained in the remainder of this section.
```

Or add a subheading above here.

Author: Added subheading, thanks

If the `_X_AMZN_TRACE_ID` environment variable is set, it SHOULD be parsed into an OpenTelemetry `Context` using
the [AWS X-Ray Propagator](../../../context/api-propagators.md). If the resulting `Context` is sampled, then this
`Context` is the parent of the function span. The environment variable will be set and the `Context` will be
sampled only if AWS X-Ray has been enabled for the Lambda function. A user can disable AWS X-Ray for the function
if this propagation is not desired.

Otherwise, for an API Gateway Proxy Request, the user's configured propagators should be applied to the HTTP
headers of the request to extract a `Context`.

## API Gateway

API Gateway allows a user to trigger a Lambda function in response to HTTP requests. It can be configured to be
a pure proxy, where the information about the original HTTP request is passed to the Lambda function, or as a
configuration for a REST API, in which case only a deserialized body payload is available.  In the case the API
gateway is configured to proxy to the Lambda function, the instrumented request handler will have access to all
the information about the HTTP request in the form of an API Gateway Proxy Request Event.

> The Lambda span name SHOULD be set to the `Resource` from the proxy request event, which corresponds to the user

Reviewer 2: ```suggestion
The Lambda span name and the [`http.route` span attribute](../http.md) SHOULD be set to the `Resource` from the proxy request event, which corresponds to the user
```

Resolves https://github.com/open-telemetry/opentelemetry-specification/pull/1442/files#r576699975

configured HTTP route instead of the function name.

[`faas.trigger`](../faas.md) MUST be set to `http`. [HTTP attributes](../http.md) SHOULD be set based on the
available information in the proxy request event.

## SQS

SQS is a message queue that triggers a Lambda function with a batch of messages. So we consider processing both
of a batch and of each individual message. The function invocation span MUST correspond to the SQS event, which
is the batch of messages. For each message, an additional span SHOULD be created to correspond with the handling
of the SQS message. Because handling of a message will be inside user business logic, not the Lambda framework,

> automatic instrumentation mechanisms without code change will often not be able to be able to instrument the

Reviewer 2: ```suggestion
automatic instrumentation mechanisms without code change will often not be able to instrument the
```

processing of the individual messages.

The span kind for both spans MUST be `CONSUMER`.

For the SQS event span, if all the messages in the event have the same event source, the name of the span MUST
be `<event source> process`. If there are multiple sources in the batch, the name MUST be
`multiple_sources <process>`. The parent MUST be the `SERVER` span corresponding to the function invocation.

For every message in the event, the message's system attributes (not message attributes, which are provided by
the user) SHOULD be checked for the key `AWSTraceHeader`. If it is present, an OpenTelemetry `Context` SHOULD be
parsed from the value of the attribute using the [AWS X-Ray Propagator](../../../context/api-propagators.md) and
added as a link to the span. This means the span may have as many links as messages in the batch.

[`faas.trigger`](../faas.md) MUST be set to `pubsub`.

> [`messaging.operation`](../messaging.md) MUST be set to `process`.

Reviewer 2: It irks me that both the span receiving the message and the spans processing the messages have the "process" operation. But I think you are right here and the "receive" operation is meant for synchronous receiving as it is currently defined.

Maybe this case would warrant a special `messaging.operation`, maybe `dispatch`? "Used for spans corresponding to a middleware component that gets spans received by a lower level component but contains no business logic for processing the messages, instead passing the message(s) on to other components which process it."

Although from what I understand, the processing would happen synchronously inside child spans of this span, or, if there is no instrumentation for the actual processing, there would be no child spans and this span is the closes equivalent to "process".

Author: Yeah I think a separate operation type could make sense for the batch case. Presuming for now this is ok to go with

Reviewer 2: CC @arminru

[`messaging.system`](../messaging.md) MUST be set to `AmazonSQS`.

> For the SQS message span, the name MUST be `<event source> process`.  The parent MUST be the `CONSUMER` span

Reviewer 2: Please add subsection headings for message spans and event span for clarity.

> # Instrumenting AWS Lambda

Reviewer 3: Please add this to the README one level above where the other semantic conventions are mentioned (preferably as a sub-item for FaaS).

**Status**: [Experimental](../../../document-status.md)

This document defines how to apply semantic conventions when instrumenting an AWS Lambda request handler. AWS
Lambda largely follows the conventions for [FaaS](../faas.md) while [HTTP](../http.md) conventions are also
applicable when handlers are for HTTP requests.

There are a variety of triggers for Lambda functions, and this document will grow over time to cover all the
use cases.

## All triggers

For all events, a span with kind `SERVER` MUST be created corresponding to the function invocation unless stated
otherwise below. Unless stated otherwise below, the name of the span MUST be set to the function name from the
Lambda `Context`.

> The following attributes SHOULD be set.

Reviewer 3: ```suggestion
The following attributes SHOULD be set:
```

- [`faas.execution`](../faas.md) - The value of the AWS Request ID, which is always available through an accessor on the Lambda `Context`

> - [`faas.id`](../../../resource/semantic_conventions/faas.md) - The value of the invocation arn for the function, which is always available through an accessor on the Lambda `Context`

Reviewer 3: ```suggestion
- [`faas.id`](../../../resource/semantic_conventions/faas.md) - The value of the invocation ARN for the function, which is always available through an accessor on the Lambda `Context`
```

If the `_X_AMZN_TRACE_ID` environment variable is set, it SHOULD be parsed into an OpenTelemetry `Context` using
the [AWS X-Ray Propagator](../../../context/api-propagators.md). If the resulting `Context` is sampled, then this
`Context` is the parent of the function span. The environment variable will be set and the `Context` will be
sampled only if AWS X-Ray has been enabled for the Lambda function. A user can disable AWS X-Ray for the function
if this propagation is not desired.

Otherwise, for an API Gateway Proxy Request, the user's configured propagators should be applied to the HTTP
headers of the request to extract a `Context`.

## API Gateway

API Gateway allows a user to trigger a Lambda function in response to HTTP requests. It can be configured to be
a pure proxy, where the information about the original HTTP request is passed to the Lambda function, or as a
configuration for a REST API, in which case only a deserialized body payload is available.  In the case the API
gateway is configured to proxy to the Lambda function, the instrumented request handler will have access to all
the information about the HTTP request in the form of an API Gateway Proxy Request Event.

> The Lambda span name and the [`http.route` span attribute](../http.md) SHOULD be set to the `Resource` from the

Reviewer 3: When reading this I briefly thought this would be about the OpenTelemetry Resource due to the capitalization and formatting. Could you link to the definition for this event here or in the previous sentence so one could look it up when reading?

This would also clear it up:
```suggestion
The Lambda span name and the [`http.route` span attribute](../http.md#http-server-semantic-conventions) SHOULD be set to the `resource` property from the
```

proxy request event, which corresponds to the user configured HTTP route instead of the function name.

[`faas.trigger`](../faas.md) MUST be set to `http`. [HTTP attributes](../http.md) SHOULD be set based on the
available information in the proxy request event.

## SQS

> SQS is a message queue that triggers a Lambda function with a batch of messages. So we consider processing both

Reviewer 3: ```suggestion
Amazon Simple Queue Service (SQS) is a message queue that triggers a Lambda function with a batch of messages. So we consider processing both
```
Since you already provide a definition/introduction here.

of a batch and of each individual message. The function invocation span MUST correspond to the SQS event, which
is the batch of messages. For each message, an additional span SHOULD be created to correspond with the handling
of the SQS message. Because handling of a message will be inside user business logic, not the Lambda framework,
automatic instrumentation mechanisms without code change will often not be able to instrument the processing of
the individual messages.

> The span kind for both spans MUST be `CONSUMER`.

Reviewer 5: If I follow this correctly, processing a single SQS event with 2 messages should yield following trace (parent - child links):
SQS event
 - SQS msg 1
 - SQS msg 2
 - spans from autoinstrumentation

Author: Ah, `both spans` is a bit misleading, reworded a bit

### SQS Event

For the SQS event span, if all the messages in the event have the same event source, the name of the span MUST
be `<event source> process`. If there are multiple sources in the batch, the name MUST be

> `multiple_sources <process>`. The parent MUST be the `SERVER` span corresponding to the function invocation.

Author: Thanks, a typo indeed.

Messaging spec is definitely not set in stone so could be worth starting a Discussion on it.

> For every message in the event, the message's system attributes (not message attributes, which are provided by

Reviewer 3: ```suggestion
For every message in the event, the _message system attributes_ (not message attributes, which are provided by
```
As it's called here: https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-message-metadata.html#sqs-message-system-attributes

Ideally also add a reference to the docs here.

Author: Thanks - didn't keep the italics since with the link it renders blue

the user) SHOULD be checked for the key `AWSTraceHeader`. If it is present, an OpenTelemetry `Context` SHOULD be
parsed from the value of the attribute using the [AWS X-Ray Propagator](../../../context/api-propagators.md) and
added as a link to the span. This means the span may have as many links as messages in the batch.

[`faas.trigger`](../faas.md) MUST be set to `pubsub`.

> [`messaging.operation`](../messaging.md) MUST be set to `process`.

Reviewer 3: ```suggestion
[`messaging.operation`](../messaging.md#messaging-attributes) MUST be set to `process`.
```
The prelude there is quite lengthy so the anchor would be helpful.

> [`messaging.system`](../messaging.md) MUST be set to `AmazonSQS`.

Reviewer 3: You could add `AmazonSQS` to the examples listed for `messaging.system` so instrumentation authors can also find it outside the Lambda context. (Eventually we should make this an open enum with suggested values like we have for `db.system`).

> [`messaging.system`](../messaging.md) MUST be set to `AmazonSQS`.

Reviewer 3: ```suggestion
[`messaging.system`](../messaging.md) MUST be set to `AmazonSQS`.
[`messaging.destination`](../messaging.md) MUST be set to `queue`.
```
Setting `messaging.destination` is required "if the message destination is either a queue or topic", see
https://github.com/open-telemetry/opentelemetry-specification/blob/v1.1.0/specification/trace/semantic_conventions/messaging.md#messaging-attributes.

### SQS Message

For the SQS message span, the name MUST be `<event source> process`.  The parent MUST be the `CONSUMER` span

> corresponding to the SQS event. The message's system attributes (not message attributes, which are provided by

Reviewer 3: ```suggestion
corresponding to the SQS event. The _message system attributes_ (not message attributes, which are provided by
```

the user) SHOULD be checked for the key `AWSTraceHeader`. If it is present, an OpenTelemetry `Context` SHOULD be
parsed from the value of the attribute using the [AWS X-Ray Propagator](../../../context/api-propagators.md) and
added as a link to the span.

[`faas.trigger`](../faas.md) MUST be set to `pubsub`.
[`messaging.operation`](../messaging.md) MUST be set to `process`.

> [`messaging.system`](../messaging.md) MUST be set to `AmazonSQS`.

Reviewer 3: ```suggestion
[`messaging.system`](../messaging.md) MUST be set to `AmazonSQS`.
[`messaging.destination`](../messaging.md) MUST be set to `queue`.
```

Other [Messaging attributes](../messaging.md) SHOULD be set based on the available information in the SQS message
event.

> Note that `AWSTraceHeader` is the only supported mechanism for propagating `Context` for SQS to prevent conflicts

Reviewer 3: Supported by whom?

Reviewer 2: I think it should be possible to use any OpenTelemetry-compatible vendor with SQS without paying for X-Ray. I'd like it more if we say something to the effect that "Instrumentations SHOULD default to using AWSTraceHeader with AWS X-Ray Propagator, but SHOULD be configurable to use any other header and propagator". If that makes sense at all.

Reviewer 2: Also related: https://github.com/open-telemetry/opentelemetry-specification/pull/1442#discussion_r576698621

Author: I meant supported by our instrumentation. `AWSTraceHeader` doesn't mention the word X-Ray :) It's how to propagate context within AWS services and by having AWS SDK instrumentation and Lambda instrumentation explicitly use the X-Amzn-Trace-Id format (which we generally incorrectly call x-ray, it's actually not really tied to x-ray), it precisely allows any vendor to use AWS asynchronous services with proper propagation. No payment for X-Ray or anything involved. While it might be possible to implement our own propagation through message attributes, I'm not confident it would be propagated between services (e.g., S3 -> SNS -> SQS is a possible flow) while I am confident that `AWSTraceHeader` will be, regardless of tracing vendor. @kubawach has done incredible testing on these propagation cases, I'm guessing for use with Splunk, not X-Ray :P

I am going to try adding some text with some of this information to allay the concerns, which are fair but I wouldn't be worried :)

Reviewer 5: I support @anuraaga here. 

With the first approach to AWS propagation (for SQS producer - consumer) I used "standard" HTTP-headers mechanism, creating a SQS message attribute (upon produce) and extracting parent from it (upon consume). While this approach worked in SQS - SQS scenario, it would fail with more complex (but generally used around the world) scenarios such as S3 - SQS or S3 - SNS - SQS. Therefore I had to switch back to using AWS feature, guaranteeing that if AWS Trace header is set during AWS SDK request, the value will be maintained and returned at the end of the chain (consume - as SQS system attribute). There was simply no other way to do it (ie without relying on AWS features).

To sum up, approach that came out of discussions and code reviews was:
- when interacting with AWS (using AWS SDK) we will enforce X-Ray propagator (which basically sets AWS trace header in an appropriate format) in order to ensure that context propagation will be maintained thru the infrastructure (which is beyond our control)
- if there is a system that conforms to AWS interface and supports AWS SDK, in order to maintain the propagation it will also need to support AWS tracing header / format

Frankly at the beginning it didn't feel right to implement propagation enforcing AWS trace format but I realised that it's just relying on how a closed system work, just as we do with instrumentation libraries, in order to support as many use cases as possible.

Reviewer 6: Is this comment addressed @arminru ?

to prevent conflicts with other sources. Notably, message attributes (user-provided, not system) are not supported -
the linked contexts are always expected to have been sent as HTTP headers of the `SQS.SendMessage` request that
the message originated from. This is a function of AWS SDK instrumentation, not Lambda instrumentation.

Using the `AWSTraceHeader` ensures that propagation will work across AWS services that may be integrated to
Lambda via SQS, for example a flow that goes through S3 -> SNS -> SQS -> Lambda. `AWSTraceHeader` is only a means
of propagating context and not tied to any particular observability backend. Notably, using it does not imply
using AWS X-Ray - any observability backend will fully function using this propagation mechanism.

[message system attributes]: https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-message-metadata.html#sqs-message-system-attributes

## Examples

### API Gateway Request Proxy (Lambda tracing passive)

Given a process C that sends an HTTP request to an API Gateway endpoint with path `/pets/{petId}` configured for
a Lambda function F:

```
Process C: | Span Client        |
--

> Function F:    | Span Function |

Reviewer 3: nit: That sounds a bit odd (what's a span client?), I'd prefer "client span" and "function span".

Author: I borrowed this pattern from messaging.md. Should I change it?

Reviewer 3: Ah I see. There we use "Prod1" and "Proc1" there as symbols but "span client" sounds like someone consuming the span and "span function" sounds like something applied to a span to me. Should be fine given the context, however.

```

| Field or Attribute | `Span Client` | `Span Function` |
|-|-|-|
| Span name | `HTTP GET` | `/pets/{petId}` |
| Parent |  | Span Client |
| SpanKind | `CLIENT` | `SERVER` |
| Status | `Ok` | `Ok` |
| `faas.execution` | | `79104EXAMPLEB723` |
| `faas.id` | | `arn:aws:lambda:us-west-2:123456789012:function:my-function` |
| `faas.trigger` | | `http` |
| `cloud.account.id` | | `12345678912` |
| `net.peer.name` | `foo.execute-api.us-east-1.amazonaws.com` |  |
| `net.peer.port` | `413` |  |
| `http.method` | `GET` | `GET` |
| `http.user_agent` | `okhttp 3.0` | `okhttp 3.0` |
| `http.url` | `https://foo.execute-api.us-east-1.amazonaws.com/pets/10` |  |
| `http.scheme` | | `https` |
| `http.host` | | `foo.execute-api.us-east-1.amazonaws.com` |
| `http.target` | | `/pets/10` |
| `http.route` | | `/pets/{petId}` |
| `http.status_code` | `200` | `200` |

### API Gateway Request Proxy (Lambda tracing active)

Active tracing in Lambda means an API Gateway span `Span APIGW` and a Lambda runtime invocation span `Span Lambda`
will be exported to AWS X-Ray. All attributes above are the same except that in this case, the parent of `APIGW`

> is `Span Client` and the parent of `Span Function` is `Span Lambda`.

Reviewer 3: The two new spans are not defined in our semantic conventions, right?
Is the Lambda span a child of the APIGW span?
You could also add a "drawing" of the relationship like:
```
Span Client --> Span APIGW --> Span Lambda --> Span Function
```
with `x --> y` = `x parent of y`

### SQS (Lambda tracing passive)

Given a process P, that sends two messages to a queue Q on SQS, and a Lambda function F, which processes both of them in one batch (Span ProcBatch) and
generates a processing span for each message separately (Spans Proc1 and Proc2).

```
Process P: | Span Prod1 | Span Prod2 |
--
Function F:                      | Span ProcBatch |
                                        | Span Proc1 |
                                               | Span Proc2 |
```

| Field or Attribute | Span Prod1 | Span Prod2 | Span ProcBatch | Span Proc1 | Span Proc2 |
|-|-|-|-|-|-|
| Span name | `Q send` | `Q send` | `Q process` | `Q process` | `Q process` |
| Parent |  |  |  | Span ProcBatch | Span ProcBatch |
| Links |  |  |  | Span Prod1 | Span Prod2 |
| SpanKind | `PRODUCER` | `PRODUCER` | `CONSUMER` | `CONSUMER` | `CONSUMER` |
| Status | `Ok` | `Ok` | `Ok` | `Ok` | `Ok` |
| `messaging.system` | `AmazonSQS` | `AmazonSQS` | `AmazonSQS` | `AmazonSQS` | `AmazonSQS` |

> | `messaging.destination` | `Q` | `Q` | | `Q` | `Q` |

Reviewer 3: ```suggestion
| `messaging.destination` | `Q` | `Q` | `Q` | `Q` | `Q` |
```
Why would you not set the Queue name here?

Author: `Q` is the queue name - I copied this pattern from messaging.md too :)

Reviewer 3: Sure, my question was why you would >not< also set `Q` for the `ProcBatch` span here.

Author: Oops - thanks for the text I understand better. Actually the batch span can't have it since the same catch can serve multiple queues. I applied the suggestion thinking I made a column-length typo heh.

Reviewer 3: Ah I see! Well in this case it happens to be the same so if the batch would be inspected at that stage it could be added but that would be optional, so either should be fine here.

Author: I added text to indicate what should happen if multiple queues. We might want to copy the text into `messaging.md` too

| `messaging.destination_kind` | `queue` | `queue` | `queue` | `queue` | `queue` |

> | `messaging.operation` |  |  | `process` | `process"` | `process` |

Reviewer 3: ```suggestion
| `messaging.operation` |  |  | `process` | `process` | `process` |
```

| `messaging.message_id` | | | | `"a1"` | `"a2"` |

> The above requires user code change to create `Span Proc1` and `Span Proc2` - in Java, the user would inherit from

Reviewer 3: ```suggestion
The above requires user code change to create `Span Proc1` and `Span Proc2`. In Java, the user would inherit from
```

> [TracingSqsMessageHandler][] instead of Lambda's standard `RequestHandler` to enable them. Otherwise the spans would
> not exist.

Reviewer 3: ```suggestion
[TracingSqsMessageHandler][] instead of Lambda's standard `RequestHandler` to enable them. Otherwise these two spans
would not exist.
```

> [TracingSqsMessageHandler]: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/aws-lambda-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambda/v1_0/TracingSqsMessageHandler.java

Reviewer 3: ```suggestion
[TracingSqsMessageHandler]: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/v1.0.1/instrumentation/aws-lambda-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambda/v1_0/TracingSqsMessageHandler.java
```

### SQS (Lambda tracing active)

Active tracing in Lambda means a Lambda runtime invocation span `Span Lambda` will be exported to X-Ray. In this

> case, all of the above is the same except `Span ProcBatch` will have a parent of `Span Lambda`.

Reviewer 3: Same as above
